### PR TITLE
Added VaryingCDF, a general parameter dependant CDF. Also added Print

### PR DESCRIPTION
**Pull request by BillyLiggins**
Added VaryingCDF, a general parameter dependant CDF. Also added Print functionality to the matrix classes.

A VaryingCDF is a CDF whose underlying generating kernels parameters
(e.g. the stdev of the Gaussian that makes the matrix) can depend on
general functions of the fit parameters.

Also added print functions to the matrix classes.

Usage:

   Gaussian gaus("gaus",0,1);
   gaus.RenameParameter("means_0","mean");
   gaus.RenameParameter("stddevs_0","std");
   VaryingCDF smearer("smearer");
   smearer.SetKernal(&gaus)
   smearer.SetDependance("std",ploy(see note));

   The above creates a VaryingCDF whose gaussian kernel has a variable
   stddev which goes like the function ploy.

   note:
   Where 'ploy' is a function that takes a vector of doubles that
   represents the bin in which you are evaluating on e.g. fitting in r,E
   ploy will take a vector = [r_i,E_j]. The order of the vector is the same
   as the order in which the axes of the fit were set.

### CODE REVIEW

**Comment by tsengj10** [comment]
Could you please add some descriptive comment to the top of the header files and example?

**Comment by BillyLiggins** [comment]
Hi @tsengj10, what do you think of the description I added? I'm open to comments because it doesn't feel correct somehow.

**Comment by tsengj10** [comment]
Thanks @BillyLiggins.  It's probably better not to refer to the pull request in the descriptive comment, since a lot of users may have difficulty finding it.  Can you include the PR description, especially the usage, in the appropriate files?

It's also helpful to have description/usage in the example script for VaryingCDF.cpp, so users know what it's for and how it's used (basic sense).

**Comment by BillyLiggins** [comment]
I added some more comments in the relevant places, it is probably not water tight but its what I have.

**Comment by tsengj10** [comment]
@terezakroupa had a chance to look at this (and #3)?

**Comment by terezakroupa** [comment]
@tsengj10 Yes, I've just been reviewing these PRs this week. Sorry for the delay!

File: examples/VaryingCDF.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+#include <set>
+#include <map>
+#include <vector>
+
+using ContainerTools::ToString;
+using ContainerTools::GetValues;
+using ContainerTools::GetKeys;
+
+//Quick function ploy(x) = a*sqrt(abs(x)) + b
```
Comment (terezakroupa) [inline examples/VaryingCDF.cpp:50]: Maybe this is a generally useful function that could be moved to src/function?

File: src/function/VaryingCDF.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * width doesn't encode this functional dependence. The VaryingCDF can take the
+ * gaussian (or any function) as it's kernel and change the parameters of this
+ * kernel by some general function dependent on the current spectrum point. 
+ *
+ * Usages:
+ *    Gaussian gaus("gaus",0,1);
+ *    gaus.RenameParameter("means_0","mean");
+ *    gaus.RenameParameter("stddevs_0","std");
+ *    VaryingCDF smearer("smearer");
+ *    smearer.SetKernal(&gaus)
```
Comment (terezakroupa) [inline src/function/VaryingCDF.h:16]: typo

File: src/function/VaryingCDF.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ *    smearer.SetDependance
```
Comment (terezakroupa) [inline src/function/VaryingCDF.h:17]: This line should not be there, I think

File: test/unit/VaryingCDFTest.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+#include <Function.h>
+#include <Gaussian.h>
+#include <math.h>
+
+
+using ContainerTools::ToString;
+using ContainerTools::GetValues;
+using ContainerTools::GetKeys;
+
+class Ploy : public Function{
```
Comment (terezakroupa) [inline test/unit/VaryingCDFTest.cpp:17]: Again, maybe this could either be moved to src/function, or even better done with the ploy from the example?

**Review by terezakroupa — APPROVED** [review]

**Review by terezakroupa — COMMENTED** [review]
Also, I found it confusing that CDF stands for conditional probability distribution (as opposed to cumulative distribution function), so maybe it's just worth mentioning that in the comments.

**Comment by terezakroupa** [comment]
This all works well, all my comments are just suggestions for which I wouldn't mind a second opinion. @tsengj10 Do you think it's worth merging this and I can PR the changes separately and other people can have a look at them? Especially as it's unreasonable to expect @BillyLiggins to do anything about them at this point.

**Comment by tsengj10** [comment]
Yes, I was thinking that as well.  Make a new PR which references this one, so it's easier for a reviewer to see how the changes are related to your comments on this one.

I'll add the merge-soon label, and then you (or someone else) can go ahead and merge in a few hours.

**Comment by terezakroupa** [comment]
Sounds good, thank you

**Comment by tsengj10** [comment]
@terezakroupa do you have a "Merge pull request" button on this page?

**Comment by terezakroupa** [comment]
Yes, I do. Will merge now

**Pull request merged**